src/cluster/mcl.cpp

```diff
@@ -0,0 +1,570 @@
+
+	task_timer timer;
+	timer.go("Computing independent components");
+	auto componentInfo = ms.getComponents();
+	vector<vector<uint32_t>> indices = get<0>(componentInfo);
+	vector<vector<Eigen::Triplet<float>>> components = get<1>(componentInfo);
+	vector<uint32_t> sort_order(components.size());
+	iota(sort_order.begin(), sort_order.end(), 0);
+	sort(sort_order.begin(), sort_order.end(), [&](uint32_t i, uint32_t j){return indices[i].size() > indices[j].size();});
+
+	uint32_t nComponents = count_if(indices.begin(), indices.end(), [](vector<uint32_t> v){ return v.size() > 0;});
+	uint32_t nComponentsLt1 = count_if(indices.begin(), indices.end(), [](vector<uint32_t> v){ return v.size() > 1;});
+	timer.finish();
+	cout << "Found " << nComponentsLt1 <<" ("<<nComponents<< " incl. singletons) disconnected components" << endl;
+
+	timer.go("Clustering components");
+	bool full = false;
+	// Note, we will access the clustering_result from several threads below and a vector does not guarantee thread-safety in these situations.
+	// Note also, that the use of the disjoint_set structure guarantees that each thread will access a different part of the clustering_result
+	uint32_t* clustering_result = new uint32_t[db->ref_header.sequences];
+	for(uint64_t iseq = 0; iseq < db->ref_header.sequences; iseq++){
+		clustering_result[iseq]  = numeric_limits<uint32_t>::max();
+	}
+
+	uint32_t nThreads = min(config.threads_, nComponentsLt1);
+	// note that the disconnected components are sorted by size
+	const uint32_t max_counter = sort_order.size();
+	float* max_sparsities = new float[nThreads];
+	float* min_sparsities = new float[nThreads];
+	const float inflation = (float) config.cluster_mcl_inflation;
+	const float expansion = (float) config.cluster_mcl_expansion;
+
+	atomic_uint n_clusters_found(0);
+	atomic_uint component_counter(0);
+	atomic_uint n_dense_calculations(0);
+	atomic_uint n_sparse_calculations(0);
+	atomic_uint nClustersEq1(0);
+
+	auto mcl_clustering = [&](const uint32_t iThr){
+		uint32_t n_dense = 0;
+		uint32_t n_sparse = 0;
+		uint32_t n_singletons = 0;
+		uint32_t cluster_id = iThr;
+		//uint32_t my_counter = component_counter.fetch_add(1, memory_order_relaxed);
+		//while(my_counter < max_counter){
+		for(uint32_t my_counter = iThr; my_counter<max_counter; my_counter += nThreads){
+			uint32_t iComponent = sort_order[my_counter];
+			vector<uint32_t> order = indices[iComponent];
+			if(order.size() > 1){
+				vector<Eigen::Triplet<float>> m = components[iComponent];
+				assert(m.size() <= order.size() * order.size());
+				float sparsity = 1.0-(1.0 * m.size()) / (order.size()*order.size());
+				max_sparsities[iThr] = max(max_sparsities[iThr], sparsity);
+				min_sparsities[iThr] = min(min_sparsities[iThr], sparsity);
+				// map back to original ids
+				vector<unordered_set<uint32_t>> list_of_sets;
+				unordered_set<uint32_t> attractors;
+
+				//TODO: a size limit for the dense matrix should control this as well
+#ifdef MCL_TIMINGS
```

Comment:
The `high_resolution_clock` was very fast in my tests, so you can also go without these ifdefs if you want.

Reply:
I have mostly removed this due to the assumption described above and to the fact that I am collecting the timings with atomic fetch_add updates